include/lin/references.hpp

```diff
@@ -1,11 +1,450 @@
+ *
+ *  @tparam C %Vector type whose traits are replicated.
+ *  @tparam D Underlying referenced type.
+ *
+ *  @param stream Underlying stream.
+ *  @param i      Anchor point row index.
+ *  @param j      Anchor point column index.
+ *
+ *  @return Instance of an internal::RowVectorStreamReference or
+ *          internal::VectorStreamReference
+ *
+ *  This serves essentially as a wrapper around the vector reference
+ *  constructors requesting a length.
+ *
+ *  Lin assertions errors will be triggered if the requested dimensions aren't
+ *  possible given the reference's traits.
+ *
+ *  @sa internal::traits
+ *
+ *  @ingroup REFERENCES
+ */
+template <class C, class D, typename = std::enable_if_t<internal::conjunction<
+    internal::has_traits<C>, internal::is_vector<C>>::value>>
+constexpr auto ref(internal::Stream<D> const &stream, size_t i, size_t j, size_t n) {
+  return internal::stream_reference_t<C, D>(stream, i, j, n);
+}
+
+/** @brief Creates a stream reference with the provided dimensions.
+ *
+ *  @tparam C %Tensor type whose traits are replicated.
+ *  @tparam D Underlying referenced type.
+ *
+ *  @param stream Underlying stream.
+ *  @param i      Anchor point row index.
+ *  @param j      Anchor point column index.
+ *  @param r      Initial row dimension.
+ *  @param c      Initial column dimension.
+ *
+ *  @return Instance of an internal::MatrixStreamReference,
+ *          internal::RowVectorStreamReference, or
+ *          internal::VectorStreamReference
+ *
+ *  This serves essentially as a wrapper around the reference constructors
+ *  using default dimensions.
+ *
+ *  Lin assertions errors will be triggered if the requested dimensions aren't
+ *  possible given the reference's traits.
+ *
+ *  @sa internal::traits
+ *
+ *  @ingroup REFERENCES
+ */
+template <class C, class D, typename =
+    std::enable_if_t<internal::has_traits<C>::value>>
+constexpr auto ref(internal::Stream<D> const &stream, size_t i, size_t j, size_t r, size_t c) {
+  return internal::stream_reference_t<C, D>(stream, i, j, r, c);
+}
+
+/** @brief Creates a stream reference of a particular column of a given tensor.
+ *
+ *  @tparam D Underlying referened type.
+ *
+ *  @param stream Underlying stream.
+ *  @param j      Index of the referenced column.
+ *
+ *  @return Instance of an internal::VectorStreamReference.
+ *
+ *  This is a convenience function and the same result can be obtained with a
+ *  call to another reference function.
+ *
+ *  The dimensions of a variably sized column are set depending on the provided
+ *  mapping's row dimension at run time.
+ *
+ *  @ingroup REFERENCES
+ */
+template <class D>
+constexpr auto ref_col(internal::Stream<D> const &stream, size_t j) {
+  typedef typename D::Traits::elem_t Elem;
+  constexpr size_t Rows = D::Traits::rows;
+  constexpr size_t MaxRows = D::Traits::max_rows;
+
+  return ref<Vector<Elem, Rows, MaxRows>>(stream, 0, j, stream.rows());
```

Comment:
what is a stream again? and why does this say .rows() when referring to cols?

I acknowledge the possibility of this being inverted intentionally but I wanted to 2x check

Reply:
A `template <class D> lin:::internal::Stream` the parent type for everything in `lin`. It defines a read only interface for tensor elements. See https://pathfinder-for-autonomous-navigation.github.io/lin/classlin_1_1internal_1_1_stream.html.

Reply:
As for the `stream.rows()`, if the stream being considered doesn't have a fixed row count we won't actually know the length of each column until runtime. The length of each column then is given by the row count of the backing type. This would happen with a type like `lin::Matrixd<0, 5, 5, 5>`.